lib/arel/middleware/chain.rb

```diff
@@ -1,31 +1,65 @@
 module Arel
   module Middleware
+    module NoOpCache
+      def self.get(key); end
+
+      def self.set(key, sql); end
+    end
+
+    class CacheAccessor
+      attr_reader :cache
+
+      def initialize(cache)
+        @cache = cache
+      end
+
+      def get(original_sql)
+        cache.get cache_key(original_sql)
+      end
+
+      def set(transformed_sql:, transformed_binds:, original_sql:, original_binds:)
+        # The order of binds was changed and therefore we can't reuse the query
+        return if transformed_binds != original_binds
+
+        cache.set(cache_key(original_sql), transformed_sql)
+      end
+
+      def cache_key(original_sql)
+        # TODO: Add the active middleware to the cache key
```

Comment:
@mvgijssel do you know if it's possible to get the whole middleware chain? The thing I like to achieve is that the cache key becomes something like:
```
SomeMiddleware::AnotherMiddleware::SELECT * FROM kerks;
```

Reply:
Something like this?

```ruby
class Foo; def call(arel, next_middleware); next_middleware(arel); end; end
class Bar; def call(arel, next_middleware); next_middleware(arel); end; end

Arel.middleware.apply([Foo, Bar]) { Arel.middleware.current.map(&:name) }
```

Reply:
Let me try!